### DISCUSSION TITLE
PSRAM Usage

### DISCUSSION THREAD

**Discussion by caco3** [jomjol/AI-on-the-edge-device#2270] [Ideas]
The Screenshots in this thread where made with the [PSRAM Usage Visualizer](https://github.com/caco3/psram-usage-visualizer).

The memory allocation are in detail shown in https://github.com/jomjol/AI-on-the-edge-device/pull/2215 and https://github.com/jomjol/AI-on-the-edge-device/issues/2200
The timely order of the allocation is visible here:
![grafik](https://user-images.githubusercontent.com/1783586/228981389-98ea8e98-5eed-4cc7-994c-da3e3e77f6f0.png)

**Comment by caco3** [comment]
Colors:
- White: Unused
- Red: Alloc
- green: free
- light red: in use

**Comment by caco3** [comment]
# v15.1.1
`dig-cont_0611_s3.tflite` (1.1 MB) and `ana-cont_1105_s2_q.tflite` (52 kB)

=> OK

![15 1 1 annotated](https://user-images.githubusercontent.com/1783586/229174790-18810f70-d0c6-47ce-9d6d-67820fee00c9.png)

[psram_15.1.1.log](https://github.com/jomjol/AI-on-the-edge-device/files/11116754/psram_15.1.1.log)

**Comment by caco3** [comment]
# v15.0.1
`dig-cont_0611_s3.tflite` (1.1 MB) and `ana-cont_1105_s2_q.tflite` (52 kB)

=> OK

![15 0 1](https://user-images.githubusercontent.com/1783586/229175040-9382b8c4-9cca-4f55-ba14-c9478a606f5d.png)

**Comment by caco3** [comment]
# v15.1.0
`dig-cont_0611_s3.tflite` (1.1 MB) and `ana-cont_1105_s2_q.tflite` (52 kB)

=> OK but shows some fragmentation's:

![grafik](https://user-images.githubusercontent.com/1783586/229175816-fa79c5cc-30b1-47d5-9335-f6bb78af641a.png)

**Comment by caco3** [comment]
# v15.1.1
`dig-cont_0611_s3.tflite` (1.1 MB) and `ana-cont_1105_s2_q.tflite` (52 kB)

=> OK

![15 1 1](https://user-images.githubusercontent.com/1783586/229175122-e41701ec-23a3-4b9c-89e9-065b198046aa.png)

**Comment by caco3** [comment]
# Branch shared-psram-objects 3e752fde7

`dig-cont_0611_s3.tflite` (1.1 MB) and `ana-cont_1105_s2_q.tflite` (52 kB)

=> malloc failed due to fragmentation:
```
I (24932) C IMG BASIS: Image (zwImage) loaded from memory: 0, 0, 0
E (24932) C IMG BASIS: Image with size 0 loaded --> reboot to be done! Check that your camera module is working and connected properly.
```
![grafik](https://user-images.githubusercontent.com/1783586/229212978-0b0074a4-17e3-4c4a-a232-e4dbe445dc06.png)

[shared-psram-objects_3e752fde7.log](https://github.com/jomjol/AI-on-the-edge-device/files/11124013/shared-psram-objects_3e752fde7.log)

**Reply by caco3** [reply]
![grafik](https://user-images.githubusercontent.com/1783586/229213448-242cc910-722d-47bb-85a1-52c89ffd13bf.png)

=> Fragmentation at line 1695 in the log file:
```
PSRAM MALLOC 0x3fbee240, 1696 <= Fragmentation, line 1695
PSRAM MALLOC 0x3f910338, 24
PSRAM MALLOC 0x3f9106f4, 16
PSRAM MALLOC 0x3f910708, 8
PSRAM FREE 0x3f910708
PSRAM FREE 0x3fbee240 <= line 1700
```

**Reply by caco3** [reply]
The problem is that `stbi_load_from_memory()` resp STBI in general additionally allocates memory (most likely again around 900 MB). How ever we do not have that much left in PSRAM in this branch.

Since we have `CONFIG_SPIRAM_USE_MALLOC` enabled, also normal mallocs can use PSRAM. This is why we see so many small allocations in there.

**Reply by caco3** [reply]
Same branch but differen PSRAM usage:
```
-CONFIG_SPIRAM_USE_MALLOC=y
+#CONFIG_SPIRAM_USE_MALLOC=y
+CONFIG_SPIRAM_USE_CAPS_ALLOC=y
```

There is a slightly reduced noise on it, but still quite some noise, but no more fragmentation.
How ever still not enough for the STBI:

![grafik](https://user-images.githubusercontent.com/1783586/229221564-b69438db-fc97-4807-8f64-ac5a329371c1.png)

**Reply by Slider0007** [reply]
> The problem is that `stbi_load_from_memory()` resp STBI in general additionally allocates memory (most likely again around 900 MB). How ever we do not have that much left in PSRAM in this branch.
> 
> Since we have `CONFIG_SPIRAM_USE_MALLOC` enabled, also normal mallocs can use PSRAM. This is why we see so many small allocations in there.
> 

Nice graphs :-)
This proves what I've already figured out and posted in PR thread: https://github.com/jomjol/AI-on-the-edge-device/pull/2215#issuecomment-1480121093

**Reply by caco3** [reply]
Using `class11_1600_s2.tflite` (350 KB) instead of `dig-cont_0611_s3.tflite` (1.1 MB):

![grafik](https://user-images.githubusercontent.com/1783586/229222647-f2a0c339-89b3-4b24-9ea2-a29c39ef9721.png)
This fails, most likely because STBI by default uses malloc which now directs to the internal RAM.

[shared-psram-objects_3e752fde7_CONFIG_SPIRAM_USE_CAPS_ALLOC_class11_1600_s2.log](https://github.com/jomjol/AI-on-the-edge-device/files/11125694/shared-psram-objects_3e752fde7_CONFIG_SPIRAM_USE_CAPS_ALLOC_class11_1600_s2.log)

**Reply by Slider0007** [reply]
> Same branch but malloc stays in internal RAM, only `heap_caps_malloc/heap_caps_calloc` goes to PSRAM:
> ```
> -CONFIG_SPIRAM_USE_MALLOC=y
> +#CONFIG_SPIRAM_USE_MALLOC=y
> +CONFIG_SPIRAM_USE_CAPS_ALLOC=y
> ```
> 
> There is a slightly reduced noise on it, but still quite some noise, but no more fragmentation.
> How ever still not enough for the STBI:
> 
> ![grafik](https://user-images.githubusercontent.com/1783586/229221564-b69438db-fc97-4807-8f64-ac5a329371c1.png)
> 
> 

With the preallocation approach and using of zwImage, 4MB is not sufficient. Therefore I switched the approach I posted and removed zwImage to have 1MB more RAM. But as posted in PR thread. STBI helper structure is disturbing and creates another fragmentation. Up to 1MB models files it's working but not 1.2MB.

**Reply by caco3** [reply]
Redirecting STBI to PSRAM can be done like this:
```
#include "esp_heap_caps.h"

#define STBI_MALLOC(sz)           heap_caps_malloc(sz, MALLOC_CAP_SPIRAM)
#define STBI_REALLOC(p,newsz)     heap_caps_realloc(p, newsz, MALLOC_CAP_SPIRAM)
#define STBI_FREE(p)              heap_caps_free(p)
```

inside `make_stb.cpp`, but it crashed sirectly at start:

```
================ Start app_main =================
I (1719) gpio: GPIO[32]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 

***ERROR*** A stack overflow in task main has been detected.
```

**Reply by Slider0007** [reply]
> Redirecting STBI to PSRAM can be done like this:
> ```
> #include "esp_heap_caps.h"
> 
> #define STBI_MALLOC(sz)           heap_caps_malloc(sz, MALLOC_CAP_SPIRAM)
> #define STBI_REALLOC(p,newsz)     heap_caps_realloc(p, newsz, MALLOC_CAP_SPIRAM)
> #define STBI_FREE(p)              heap_caps_free(p)
> ```
> 
> inside `make_stb.cpp`, but it crashed sirectly at start:
> 
> ```
> ================ Start app_main =================
> I (1719) gpio: GPIO[32]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
> 
> ***ERROR*** A stack overflow in task main has been detected.
> ```
> 
> 
> 

But it is already allocating in PSRAM. Why do you want to manually change that?

**Reply by caco3** [reply]
@Slider0007 I guess we would now need to manually move additional stuff to PSRAM.

**Reply by caco3** [reply]
> With the preallocation approach and using of zwImage, 4MB is not sufficient. Therefore I switched the approach I posted and removed zwImage to have 1MB more RAM.

Can you point me to this specific change?
Would we then directly have 1 MB more in PSRAM? Then we should be save, IMO.
If needed, we can extend STBI to also use static PSRAM like I do it for the model and tensor_arena!

**Reply by Slider0007** [reply]
Some kilobytes internal we can grab by reducing task sizes. I did some tests before you started with the PSRAM Para but that never get merged because it was way less than we got with wifi stack approaches. But now I think it makes sense to grab that. We can get maybe 6kB internal but not more. All other things would be quite cumbersome to move manually...

**Reply by caco3** [reply]
> But it is already allocating in PSRAM. Why do you want to manually change that?

STBI by default uses malloc, see https://github.com/jomjol/AI-on-the-edge-device/blob/shared-psram-objects/code/components/jomjol_image_proc/stb_image.h#L627

**Reply by caco3** [reply]
Would it work to move the task stacks in general to PSRAM?
That would be static as well.

https://docs.espressif.com/projects/esp-idf/en/release-v3.1/api-reference/kconfig.html#envvar-CONFIG_SPIRAM_ALLOW_STACK_EXTERNAL_MEMORY

**Reply by Slider0007** [reply]
> > With the preallocation approach and using of zwImage, 4MB is not sufficient. Therefore I switched the approach I posted and removed zwImage to have 1MB more RAM.
> 
> Can you point me to this specific change? Would we then directly have 1 MB more in PSRAM? Then we should be save, IMO. If needed, we can extend STBI to also use static PSRAM like I do it for the model and tensor_arena!

All my infos I posted in the PR thread. There is also the link to my approach: https://github.com/jomjol/AI-on-the-edge-device/pull/2215#issuecomment-1480121093

**Reply by Slider0007** [reply]
> > But it is already allocating in PSRAM. Why do you want to manually change that?
> 
> STBI by default uses malloc, see https://github.com/jomjol/AI-on-the-edge-device/blob/shared-psram-objects/code/components/jomjol_image_proc/stb_image.h#L627

malloc uses PSRAM for larger allocations than 16kB. There all? / most relevant of the STBI stuff is already in PSRAM.

**Reply by Slider0007** [reply]
> Would it work to move the task stacks in general to PSRAM? That would be static as well.
> 
> https://docs.espressif.com/projects/esp-idf/en/release-v3.1/api-reference/kconfig.html#envvar-CONFIG_SPIRAM_ALLOW_STACK_EXTERNAL_MEMORY

This is what we discussed already some weeks ago.I did some tests and failed because you have to be careful what functions you can call from PSRAM stack. Also for rev. 1 boards it's most likely not useable because of PSRAM bug...

**Reply by caco3** [reply]
> malloc uses PSRAM for larger allocations than 16kB. There all? / most relevant of the STBI stuff is already in PSRAM.

You mean `CONFIG_SPIRAM_MALLOC_ALWAYSINTERNAL`, right?
But it then looks like as if it does not work correctly, because then we should not have that much small allocations in PSRAM, right? (Except what we put there by ourself, eg. MQTT outbox)

**Reply by Slider0007** [reply]
> > malloc uses PSRAM for larger allocations than 16kB. There all? / most relevant of the STBI stuff is already in PSRAM.
> 
> You mean `CONFIG_SPIRAM_MALLOC_ALWAYSINTERNAL`, right? But it then looks like as if it does not work correctly, because then we should not have that much small allocations in PSRAM, right? (Except what we put there by ourself, eg. MQTT outbox)

Yes. But I'm not aware which other rules for the malloc selection are exisitng. Maybe the libs are not included 100% in this setting. Whenver you switch off the wifi stack you do not have fragmentation. This means most of allocations seems to respect this setting but not wifi stack allocations. Otherwise you also would have fragmentation before experimenting with wifi stack.

**Reply by caco3** [reply]
I was not aware that it is mainly the wifi stack, but it does not surprise me.

Looking on https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html#config-esp-wifi-static-rx-buffer-num

It seems we can screw between static and dynamic buffers:
![grafik](https://user-images.githubusercontent.com/1783586/229236868-cc672244-405a-472d-a81f-570f08ca7ecc.png)

Maybe we should have a look there.

**Reply by Slider0007** [reply]
I also played already with these settings. but without any luck. Maybe you have more luck...

> I was not aware that it is mainly the wifi stack, but it does not surprise me.
> 
> Looking on https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html#config-esp-wifi-static-rx-buffer-num
> 
> It seems we can screw between static and dynamic buffers: ![grafik](https://user-images.githubusercontent.com/1783586/229236868-cc672244-405a-472d-a81f-570f08ca7ecc.png)
> 
> Maybe we should have a look there.

I also played already with these settings, but without any luck. Maybe you have more luck.
